-- Return statement
-- ----------------


[case testReturnValue]
import typing
def f() -> 'A':
    return A()
    return B()
def g() -> 'B':
    return A()
class A:
    pass
class B:
    pass
[out]
main: In function "f":
main, line 4: Incompatible return value type: expected __main__.A, got __main__.B
main: In function "g":
main, line 6: Incompatible return value type: expected __main__.B, got __main__.A

[case testReturnSubtype]
import typing
def f() -> 'B':
    return A()
def g() -> 'A':
    return B()
class A:
    pass
class B(A):
    pass
[out]
main: In function "f":
main, line 3: Incompatible return value type: expected __main__.B, got __main__.A

[case testReturnWithoutAValue]
import typing
def f() -> 'A':
    return
def g() -> None:
    return
class A:
    pass
[out]
main: In function "f":
main, line 3: Return value expected

[case testReturnNoneInFunctionReturningNone]
import typing
def f() -> None:
    return None
def g() -> None:
    return f()  # E: No return value expected
[out]
main: In function "g":


-- If statement
-- ------------


[case testIfStatement]

a = None # type: A
b = None # type: bool
if a:
    a = b # Fail
elif a:
    a = b # Fail
elif a:
    a = b # Fail
else:
    a = b # Fail
if b:
    pass
elif b:
    pass
if b:
    pass

class A: pass
[builtins fixtures/bool.py]
[out]
main, line 5: Incompatible types in assignment (expression has type "bool", variable has type "A")
main, line 7: Incompatible types in assignment (expression has type "bool", variable has type "A")
main, line 9: Incompatible types in assignment (expression has type "bool", variable has type "A")
main, line 11: Incompatible types in assignment (expression has type "bool", variable has type "A")


-- Loops
-- -----


[case testWhileStatement]

a = None # type: A
b = None # type: bool
while a:
    a = b    # Fail
else:
    a = b    # Fail
while b:
    b = b

class A: pass
[builtins fixtures/bool.py]
[out]
main, line 5: Incompatible types in assignment (expression has type "bool", variable has type "A")
main, line 7: Incompatible types in assignment (expression has type "bool", variable has type "A")

[case testForStatement]

a = None # type: A
b = None # type: object
for a in [A()]:
    a = b    # Fail
else:
    a = b    # Fail

class A: pass
[builtins fixtures/list.py]
[out]
main, line 5: Incompatible types in assignment (expression has type "object", variable has type "A")
main, line 7: Incompatible types in assignment (expression has type "object", variable has type "A")

[case testBreakStatement]
import typing
while None:
    break
[builtins fixtures/bool.py]
[out]

[case testContinueStatement]
import typing
while None:
    continue
[builtins fixtures/bool.py]
[out]


-- Operator assignment
-- -------------------


[case testPlusAssign]

a, b, c = None, None, None # type: (A, B, C)
a += b   # Fail
b += a   # Fail
c += a   # Fail
a += c

class A:
    def __add__(self, x: 'C') -> 'A': pass

class B:
    def __add__(self, x: A) -> 'C': pass

class C: pass
[out]
main, line 3: Unsupported operand types for + ("A" and "B")
main, line 4: Result type of + incompatible in assignment
main, line 5: Unsupported left operand type for + ("C")

[case testMinusAssign]

a, b, c = None, None, None # type: (A, B, C)
a -= b   # Fail
b -= a   # Fail
c -= a   # Fail
a -= c

class A:
    def __sub__(self, x: 'C') -> 'A': pass

class B:
    def __sub__(self, x: A) -> 'C': pass

class C: pass
[out]
main, line 3: Unsupported operand types for - ("A" and "B")
main, line 4: Result type of - incompatible in assignment
main, line 5: Unsupported left operand type for - ("C")

[case testMulAssign]

a, c = None, None # type: (A, C)
a *= a   # Fail
c *= a   # Fail
a *= c

class A:
    def __mul__(self, x: 'C') -> 'A': pass

class C: pass
[out]
main, line 3: Unsupported operand types for * ("A" and "A")
main, line 4: Unsupported left operand type for * ("C")

[case testDivAssign]

a, c = None, None # type: (A, C)
a /= a   # Fail
c /= a   # Fail
a /= c

class A:
    def __truediv__(self, x: 'C') -> 'A': pass

class C: pass
[out]
main, line 3: Unsupported operand types for / ("A" and "A")
main, line 4: Unsupported left operand type for / ("C")

[case testPowAssign]

a, c = None, None # type: (A, C)
a **= a   # Fail
c **= a   # Fail
a **= c

class A:
    def __pow__(self, x: 'C') -> 'A': pass

class C: pass
[out]
main, line 3: Unsupported operand types for ** ("A" and "A")
main, line 4: Unsupported left operand type for ** ("C")

[case testSubtypesInOperatorAssignment]

a, b = None, None # type: (A, B)
b += b
b += a
a += b

class A:
    def __add__(self, x: 'A') -> 'B': pass

class B(A): pass
[out]

[case testAdditionalOperatorsInOpAssign]

a, c = None, None # type: (A, C)
a &= a  # Fail
a >>= a # Fail
a //= a # Fail
a &= c
a >>= c
a //= c
class A:
    def __and__(self, x: 'C') -> 'A': pass
    def __rshift__(self, x: 'C') -> 'A': pass
    def __floordiv__(self, x: 'C') -> 'A': pass
class C: pass
[out]
main, line 3: Unsupported operand types for & ("A" and "A")
main, line 4: Unsupported operand types for >> ("A" and "A")
main, line 5: Unsupported operand types for // ("A" and "A")

[case testInplaceOperatorMethods]
import typing
class A:
    def __iadd__(self, x: int) -> 'A': pass
    def __imul__(self, x: str) -> 'A': pass
a = A()
a += 1
a *= ''
a += '' # E: Argument 1 to "__iadd__" of "A" has incompatible type "str"; expected "int"
a *= 1  # E: Argument 1 to "__imul__" of "A" has incompatible type "int"; expected "str"


-- Assert statement
-- ----------------


[case testAssert]
import typing
assert None + None # Fail
assert None
[out]
main, line 2: Unsupported left operand type for + (None)


-- Exception handling
-- ------------------


[case testRaiseStatement]

e = None # type: BaseException
f = None # type: MyError
a = None # type: A
raise a # Fail
raise e
raise f
class A: pass
class MyError(BaseException): pass
[builtins fixtures/exception.py]
[out]
main, line 5: Exception must be derived from BaseException

[case testRaiseClassobject]
import typing
class A: pass
class MyError(BaseException): pass
def f(): pass
raise BaseException
raise MyError
raise A # E: Exception must be derived from BaseException
raise object # E: Exception must be derived from BaseException
raise f # E: Exception must be derived from BaseException
[builtins fixtures/exception.py]

[case testRaiseFromStatement]

e = None # type: BaseException
f = None # type: MyError
a = None # type: A
raise e from a # E: Exception must be derived from BaseException
raise e from e
raise e from f
class A: pass
class MyError(BaseException): pass
[builtins fixtures/exception.py]

[case testRaiseFromClassobject]
import typing
class A: pass
class MyError(BaseException): pass
def f(): pass
raise BaseException from BaseException
raise BaseException from MyError
raise BaseException from A # E: Exception must be derived from BaseException
raise BaseException from object # E: Exception must be derived from BaseException
raise BaseException from f # E: Exception must be derived from BaseException
[builtins fixtures/exception.py]

[case testTryFinallyStatement]
import typing
try:
    b = object() # type: A # Fail
finally:
    c = object() # type: A # Fail
class A: pass
[out]
main, line 3: Incompatible types in assignment (expression has type "object", variable has type "A")
main, line 5: Incompatible types in assignment (expression has type "object", variable has type "A")

[case testSimpleTryExcept]

try:
  pass
except BaseException as e:
  a, o = None, None # type: (BaseException, object)
  e = a
  e = o # Fail
class A: pass
class B: pass
[builtins fixtures/exception.py]
[out]
main, line 7: Incompatible types in assignment (expression has type "object", variable has type "BaseException")

[case testTypeErrorInBlock]

while object:
  x = None # type: A
  x = object()
  x = B()
class A: pass
class B: pass
[out]
main, line 4: Incompatible types in assignment (expression has type "object", variable has type "A")
main, line 5: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testTypeErrorInvolvingBaseException]

x, a = None, None # type: (BaseException, A)
a = BaseException()  # Fail
a = object()         # Fail
x = object()         # Fail
x = A()              # Fail
x = BaseException()
class A: pass
[builtins fixtures/exception.py]
[out]
main, line 3: Incompatible types in assignment (expression has type "BaseException", variable has type "A")
main, line 4: Incompatible types in assignment (expression has type "object", variable has type "A")
main, line 5: Incompatible types in assignment (expression has type "object", variable has type "BaseException")
main, line 6: Incompatible types in assignment (expression has type "A", variable has type "BaseException")

[case testSimpleTryExcept2]
import typing
try:
  pass
except BaseException as e:
  e = object() # Fail
  e = BaseException()
[builtins fixtures/exception.py]
[out]
main, line 5: Incompatible types in assignment (expression has type "object", variable has type "BaseException")

[case testBaseClassAsExceptionTypeInExcept]
import typing
try:
  pass
except Err as e:
  e = BaseException() # Fail
  e = Err()
class Err(BaseException): pass
[builtins fixtures/exception.py]
[out]
main, line 5: Incompatible types in assignment (expression has type "BaseException", variable has type "Err")

[case testMultipleExceptHandlers]
import typing
try:
    pass
except BaseException as e:
    pass
except Err as f:
    f = BaseException() # Fail
    f = Err()
class Err(BaseException): pass
[builtins fixtures/exception.py]
[out]
main, line 7: Incompatible types in assignment (expression has type "BaseException", variable has type "Err")

[case testTryExceptStatement]
import typing
try:
    a = B() # type: A       # Fail
except BaseException as e:
    e = A()             # Fail
    e = Err()
except Err as f:
    f = BaseException() # Fail
    f = Err()
class A: pass
class B: pass
class Err(BaseException): pass
[builtins fixtures/exception.py]
[out]
main, line 3: Incompatible types in assignment (expression has type "B", variable has type "A")
main, line 5: Incompatible types in assignment (expression has type "A", variable has type "BaseException")
main, line 8: Incompatible types in assignment (expression has type "BaseException", variable has type "Err")

[case testTryExceptWithinFunction]
import typing
def f() -> None:
  try: pass
  except BaseException as e:
    e = object() # Fail
    e = BaseException()
  except Err as f:
    f = BaseException() # Fail
    f = Err()
class Err(BaseException): pass
[builtins fixtures/exception.py]
[out]
main: In function "f":
main, line 5: Incompatible types in assignment (expression has type "object", variable has type "BaseException")
main, line 8: Incompatible types in assignment (expression has type "BaseException", variable has type "Err")

[case testTryWithElse]
import typing
try: pass
except BaseException: pass
else:
  object(None) # E: Too many arguments for "object"
[builtins fixtures/exception.py]

[case testExceptWithoutType]
import typing
try:
    -None # E: Unsupported operand type for unary - (None)
except:
    ~None # E: Unsupported operand type for ~ (None)
[builtins fixtures/exception.py]

[case testRaiseWithoutArgument]
import typing
try:
    None
except:
    raise
[builtins fixtures/exception.py]

[case testExceptWithMultipleTypes]
import typing
class E1(BaseException): pass
class E2(E1): pass
try:
    pass
except (E1, E2): pass
except (E1, object): pass # E: Exception type must be derived from BaseException
except (object, E2): pass # E: Exception type must be derived from BaseException
[builtins fixtures/exception.py]

[case testExceptWithMultipleTypes2]
import typing
class E1(BaseException): pass
class E2(E1): pass
try:
    pass
except (E1, E2) as e1:
    x = e1 # type: E1
    y = e1 # type: E2 # E: Incompatible types in assignment (expression has type "E1", variable has type "E2")
except (E2, E1) as e2:
    a = e2 # type: E1
    b = e2 # type: E2 # E: Incompatible types in assignment (expression has type "E1", variable has type "E2")
[builtins fixtures/exception.py]

[case testReuseTryExceptionVariable]
import typing
class E1(BaseException): pass
class E2(BaseException): pass
try: pass
except E1 as e: pass
try: pass
except E1 as e: pass
try: pass
except E2 as e: pass # E: Incompatible types in assignment (expression has type "E2", variable has type "E1")
[builtins fixtures/exception.py]

[case testArbitraryExpressionAsExceptionType]
import typing
a = BaseException
try: pass
except a as b:
    b = BaseException()
    b = object() # E: Incompatible types in assignment (expression has type "object", variable has type "BaseException")
[builtins fixtures/exception.py]

[case testInvalidExceptionCallable]
import typing
def exc() -> BaseException: pass
try: pass
except exc as e: pass             # E: Exception type must be derived from BaseException
except BaseException() as b: pass # E: Exception type must be derived from BaseException
[builtins fixtures/exception.py]

[case testOverloadedExceptionType]
from typing import overload
class E(BaseException):
    @overload
    def __init__(self) -> None: pass
    @overload
    def __init__(self, x) -> None: pass
try:
    pass
except E as e:
    e = E()
    e = BaseException() # E: Incompatible types in assignment (expression has type "BaseException", variable has type "E")
[builtins fixtures/exception.py]


-- Del statement
-- -------------


[case testDelStmtWithIndex]

a, b = None, None # type: (A, B)
del b[a]
del b[b] # E: Argument 1 to "__delitem__" of "B" has incompatible type "B"; expected "A"
del a[a] # E: "A" has no attribute "__delitem__"
del a[b] # E: "A" has no attribute "__delitem__"
class B:
  def __delitem__(self, index: 'A'): pass
class A: pass

[case testDelStmtWithAttribute]
import typing
class A:
    def f(self): pass
    x = 0
a = A()
del a.f
del a.x
del a.z # E: "A" has no attribute "z"


-- Yield statement
-- ---------------


[case testSimpleYield]
from typing import Iterator
def f() -> Iterator[int]:
    yield 1
    yield '' # E: Incompatible types in yield (actual type "str", expected type "int")
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testYieldInFunctionReturningAny]
from typing import Any
def f() -> Any:
    yield object()
[out]

[case testYieldInFunctionReturningFunction]
from typing import Callable
def f() -> Callable[[], None]:
    yield object() # E: Iterator or Generator function return type expected for "yield"
[out]
main: In function "f":

[case testYieldInDynamicallyTypedFunction]
import typing
def f():
    yield f

[case testWithInvalidInstanceReturnType]
import typing
def f() -> int:
    yield 1 # E: Iterator or Generator function return type expected for "yield"
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testTypeInferenceContextAndYield]
from typing import List, Iterator
def f() -> 'Iterator[List[int]]':
    yield []
    yield [object()] # E: List item 0 has incompatible type "object"
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testYieldAndReturnWithoutValue]
from typing import Iterator
def f() -> Iterator[int]:
    yield 1
    return
[builtins fixtures/for.py]

[case testYieldAndReturnWithValue]
from typing import Iterator
def f() -> Iterator[int]:
    yield 1
    return 1  # E: No return value expected
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testYieldWithNoValue]
from typing import Iterator
def f() -> Iterator[None]:
    yield
[builtins fixtures/for.py]

[case testGeneratorYieldType]
from typing import Generator

class A: pass
class B(A): pass
class C(B): pass

def f() -> Generator[B, None, None]:
    yield C()
    yield A()  # E: Incompatible types in yield (actual type "A", expected type "B")
[out]
main: In function "f":

[case testGeneratorReturnType]
from typing import Generator
def f() -> Generator[int, None, None]:
    yield 1
    return 1  # Fail

class A: pass
class B(A): pass

def g() -> Generator[int, None, B]:
    yield 1
    return A()  # Fail
def h() -> Generator[int, None, A]:
    yield 1
    return B()  # OK
[out]
main: In function "f":
main, line 4: No return value expected
main: In function "g":
main, line 11: Incompatible return value type: expected __main__.B, got __main__.A

[case testGeneratorSendType]
from typing import Generator
class A: pass
class B(A): pass
class C(B): pass

def f() -> Generator[int, B, None]:
    yield 1

ff = f()
ff.send(A())  # E: Argument 1 to "send" of "Generator" has incompatible type "A"; expected "B"
ff.send(C())
ff = ff.send(B())  # E: Incompatible types in assignment (expression has type "int", variable has type Generator[int, B, None])


-- Yield from generator
-- --------------------


[case testYieldFromGenerator]
from typing import Generator
class A: pass
class B: pass
class C: pass

f = None  # type: Generator[A, B, C]

def g() -> Generator[A, B, C]:
  yield from f

def h() -> Generator[B, B, C]:
  yield from f

def i() -> Generator[A, A, C]:
  yield from f

def j() -> Generator[A, B, A]:
  yield from f
[out]
main: In function "h":
main, line 12: Incompatible types in "yield from" (actual type Generator[A, B, C], expected type Generator[B, B, C])
main: In function "i":
main, line 15: Incompatible types in "yield from" (actual type Generator[A, B, C], expected type Generator[A, A, C])
main: In function "j":
main, line 18: Incompatible types in "yield from" (actual type Generator[A, B, C], expected type Generator[A, B, A])


-- Yield from Iterables
-- --------------------

[case testSimpleYieldFromWithIterator]
from typing import Iterator
def g() -> Iterator[str]:
    yield '42'
def h() -> Iterator[int]:
    yield 42
def f() -> Iterator[str]:
    yield from g()
    yield from h()  # E: Incompatible types in "yield from" (actual type "int", expected type "str")
[out]
main: In function "f":

[case testYieldFromAppliedToAny]
from typing import Any
def g() -> Any:
    yield object()
def f() -> Any:
    yield from g()
[out]

[case testYieldFromInFunctionReturningFunction]
from typing import Iterator, Callable
def g() -> Iterator[int]:
    yield 42
def f() -> Callable[[], None]:
    yield from g()  # E: Iterable or Generator function return type expected for "yield from"
[out]
main: In function "f":

[case testYieldFromNotIterableReturnType]
from typing import Iterator
def g() -> Iterator[int]:
    yield 42
def f() -> int:
    yield from g()  # E: Iterable or Generator function return type expected for "yield from"
[out]
main: In function "f":

[case testYieldFromNotAppliedIterator]
from typing import Iterator
def g() -> int:
    return 42
def f() -> Iterator[int]:
    yield from g()  # E: "yield from" can't be applied to "int"
[out]
main: In function "f":

[case testYieldFromCheckIncompatibleTypesTwoIterables]
from typing import List, Iterator
def g() -> Iterator[List[int]]:
    yield [2, 3, 4]
def f() -> Iterator[List[int]]:
    yield from g()
    yield from [1, 2, 3]  # E: Incompatible types in "yield from" (actual type "int", expected type List[int])
[builtins fixtures/for.py]
[out]
main: In function "f":

[case testYieldFromNotAppliedToNothing]
def h():
    yield from  # E: Parse error before end of line
[out]
main: In function "h":

[case testYieldFromAndYieldTogether]
from typing import Iterator
def f() -> Iterator[str]:
    yield "g1 ham"
    yield from g()
    yield "g1 eggs"
def g() -> Iterator[str]:
    yield "g2 spam"
    yield "g2 more spam"
[out]


-- With statement
-- --------------


[case testSimpleWith]
import typing
class A:
    def __enter__(self) -> None: pass
    def __exit__(self, x, y, z) -> None: pass
with A():
    object(A) # E: Too many arguments for "object"

[case testWithStmtAndInvalidExit]
import typing
class A:
    def __enter__(self) -> None: pass
    def __exit__(self, x, y) -> None: pass
with A(): # E: Too many arguments for "__exit__" of "A"
    pass

[case testWithStmtAndMissingExit]
import typing
class A:
    def __enter__(self) -> None: pass
with A(): # E: "A" has no attribute "__exit__"
    pass

[case testWithStmtAndInvalidEnter]
import typing
class A:
    def __enter__(self, x) -> None: pass
    def __exit__(self, x, y, z) -> None: pass
with A(): # E: Too few arguments for "__enter__" of "A"
    pass

[case testWithStmtAndMissingEnter]
import typing
class A:
    def __exit__(self, x, y, z) -> None: pass
with A(): # E: "A" has no attribute "__enter__"
    pass

[case testWithStmtAndMultipleExprs]
import typing
class A:
    def __enter__(self) -> None: pass
    def __exit__(self, x, y, z) -> None: pass
class B:
    def __enter__(self) -> None: pass
with A(), B(): # E: "B" has no attribute "__exit__"
    pass
with B(), A(): # E: "B" has no attribute "__exit__"
    pass

[case testWithStmtAndResult]
import typing
class B: pass
class A:
    def __enter__(self) -> B: pass
    def __exit__(self, x, y, z): pass
with A() as b:
    b = B()
    b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testWithStmtAndMultipleResults]
from typing import TypeVar, Generic
t = TypeVar('t')
class B: pass
class C: pass
class A(Generic[t]):
    def __enter__(self) -> t: pass
    def __exit__(self, x, y, z): pass
a_b = A() # type: A[B]
a_c = A() # type: A[C]
with a_b as b, a_c as c:
    b = B()
    c = C()
    b = c # E: Incompatible types in assignment (expression has type "C", variable has type "B")
    c = b # E: Incompatible types in assignment (expression has type "B", variable has type "C")

[case testWithStmtAndComplexTarget]
from typing import Tuple
class A:
    def __enter__(self) -> Tuple[int, str]: pass
    def __exit__(self, x, y, z): pass
with A() as (a, b):
    a = 1
    b = ''
    a = b # E: Incompatible types in assignment (expression has type "str", variable has type "int")
[builtins fixtures/tuple.py]


-- Chained assignment
-- ------------------


[case testChainedAssignment]
import typing
class A: pass
class B: pass
x = y = A()
x = A()
y = A()
x = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
y = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testChainedAssignment2]
import typing
def f() -> None:
    x = 1
    y = 'x'
    x = y = 'x' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
    x = y = 1   # E: Incompatible types in assignment (expression has type "int", variable has type "str")
[builtins fixtures/primitives.py]
[out]
main: In function "f":

[case testChainedAssignmentWithType]

x = y = None # type: int
x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
y = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
x = 1
y = 1


-- Star assignment
-- ---------------


[case testAssignListToStarExpr]
from typing import List
bs, cs = None, None # type: List[A], List[B]
*bs, b = bs
*bs, c = cs  # E: Incompatible types in assignment (expression has type List[B], variable has type List[A])
*ns, c = cs
nc = cs

class A: pass
class B: pass
[builtins fixtures/list.py]


-- Type aliases
-- ------------


[case testSimpleTypeAlias]
import typing
foo = int
def f(x: foo) -> None: pass
f(1)
f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"

[case testTypeAliasDefinedInAModule]
import typing
import m
def f(x: m.foo) -> None: pass
f(1)
f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"
[file m.py]
import typing
foo = int

[case testTypeAliasDefinedInAModule2]
import typing
from m import foo
def f(x: foo) -> None: pass
f(1)
f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"
[file m.py]
import typing
foo = int


-- nonlocal and global
-- -------------------


[case testTypeOfGlobalUsed]
import typing
g = A()
def f() -> None:
    global g
    g = B()

class A(): pass
class B(): pass
[out]
main: In function "f":
main, line 5: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testTypeOfNonlocalUsed]
import typing
def f() -> None:
    a = A()
    def g() -> None:
        nonlocal a
        a = B()

class A(): pass
class B(): pass
[out]
main: In function "g":
main, line 6: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testTypeOfOuterMostNonlocalUsed]
import typing
def f() -> None:
    a = A()
    def g() -> None:
        a = B()
        def h() -> None:
            nonlocal a
            a = A()
            a = B()

class A(): pass
class B(): pass
[out]
main: In function "h":
main, line 8: Incompatible types in assignment (expression has type "A", variable has type "B")
