-- Test cases for the type checker related to functions, function types and
-- calls.

-- See also check-varargs.test.


-- Callable type basics
-- --------------------


[case testCallingVariableWithFunctionType]
from typing import Undefined, Callable
f = Undefined # type: Callable[[A], B]
a, b = Undefined, Undefined # type: (A, B)
a = f(a)    # E: Incompatible types in assignment (expression has type "B", variable has type "A")
b = f(b)    # E: Argument 1 has incompatible type "B"; expected "A"
b = f()     # E: Too few arguments
b = f(a, a) # E: Too many arguments
b = f(a)

class A: pass
class B: pass

[case testSubtypingFunctionTypes]
from typing import Undefined, Callable

class A: pass
class B(A): pass

f = Undefined(Callable[[B], A])
g = Undefined(Callable[[A], A])  # subtype of f
h = Undefined(Callable[[B], B])  # subtype of f
g = h  # E: Incompatible types in assignment (expression has type Callable[[B], B], variable has type Callable[[A], A])
h = f  # E: Incompatible types in assignment (expression has type Callable[[B], A], variable has type Callable[[B], B])
h = g  # E: Incompatible types in assignment (expression has type Callable[[A], A], variable has type Callable[[B], B])
g = f  # E: Incompatible types in assignment (expression has type Callable[[B], A], variable has type Callable[[A], A])
f = g
f = h
f = f
g = g
h = h

[case testFunctionTypeCompatibilityWithOtherTypes]
from typing import Undefined, Callable
f = Undefined # type: Callable[[], None]
a, o = Undefined, Undefined # type: (A, object)
a = f   # E: Incompatible types in assignment (expression has type Callable[[], None], variable has type "A")
f = a   # E: Incompatible types in assignment (expression has type "A", variable has type Callable[[], None])
f = o   # E: Incompatible types in assignment (expression has type "object", variable has type Callable[[], None])
f = f() # E: Function does not return a value

f = f
f = None
o = f

class A: pass

[case testFunctionSubtypingWithVoid]
from typing import Undefined, Callable
f = Undefined # type: Callable[[], None]
g = Undefined # type: Callable[[], object]
f = g  # E: Incompatible types in assignment (expression has type Callable[[], object], variable has type Callable[[], None])
g = f  # E: Incompatible types in assignment (expression has type Callable[[], None], variable has type Callable[[], object])

f = f
g = g

[case testFunctionSubtypingWithMultipleArgs]
from typing import Undefined, Callable
f = Undefined # type: Callable[[A, A], None]
g = Undefined # type: Callable[[A, B], None]
h = Undefined # type: Callable[[B, B], None]
f = g  # E: Incompatible types in assignment (expression has type Callable[[A, B], None], variable has type Callable[[A, A], None])
f = h  # E: Incompatible types in assignment (expression has type Callable[[B, B], None], variable has type Callable[[A, A], None])
g = h  # E: Incompatible types in assignment (expression has type Callable[[B, B], None], variable has type Callable[[A, B], None])
g = f
h = f
h = g
f = f
g = g
h = h

class A: pass
class B(A): pass

[case testFunctionTypesWithDifferentArgumentCounts]
from typing import Undefined, Callable
f = Undefined # type: Callable[[], None]
g = Undefined # type: Callable[[A], None]
h = Undefined # type: Callable[[A, A], None]

f = g   # E: Incompatible types in assignment (expression has type Callable[[A], None], variable has type Callable[[], None])
f = h   # E: Incompatible types in assignment (expression has type Callable[[A, A], None], variable has type Callable[[], None])
h = f   # E: Incompatible types in assignment (expression has type Callable[[], None], variable has type Callable[[A, A], None])
h = g   # E: Incompatible types in assignment (expression has type Callable[[A], None], variable has type Callable[[A, A], None])

f = f
g = g
h = h

class A: pass
[out]

[case testCompatibilityOfSimpleTypeObjectWithStdType]
from typing import Undefined
t = Undefined # type: type
a = Undefined # type: A

a = A # E: Incompatible types in assignment (expression has type "A" (type object), variable has type "A")
t = f # E: Incompatible types in assignment (expression has type Callable[[], None], variable has type "type")
t = A

class A:
    def __init__(self, a: 'A') -> None: pass

def f() -> None: pass

[case testFunctionTypesWithOverloads]
from typing import Undefined, Callable, overload
f = Undefined # type: Callable[[AA], A]
g = Undefined # type: Callable[[B], B]
h = Undefined # type: Callable[[A], AA]

h = i  # E: Incompatible types in assignment (expression has type overloaded function, variable has type Callable[[A], AA])
f = j

f = i
g = i
g = j

class A: pass
class AA(A): pass

class B: pass

@overload
def i(x: AA) -> A:
    pass
@overload
def i(x: B) -> B:
    pass

@overload
def j(x: B) -> B:
    pass
@overload
def j(x: A) -> AA:
    pass

[case testOverloadWithThreeItems]
from typing import Undefined, Callable, overload
g1 = Undefined # type: Callable[[A], A]
g2 = Undefined # type: Callable[[B], B]
g3 = Undefined # type: Callable[[C], C]
g4 = Undefined # type: Callable[[A], B]
a, b, c = Undefined, Undefined, Undefined # type: (A, B, C)

b = f(a)  # E: Incompatible types in assignment (expression has type "A", variable has type "B")
a = f(b)  # E: Incompatible types in assignment (expression has type "B", variable has type "A")
b = f(c)  # E: Incompatible types in assignment (expression has type "C", variable has type "B")
g4 = f    # E: Incompatible types in assignment (expression has type overloaded function, variable has type Callable[[A], B])

g1 = f
g2 = f
g3 = f
a = f(a)
b = f(b)
c = f(c)

class A: pass
class B: pass
class C: pass

@overload
def f(x: A) -> A: pass
@overload
def f(x: B) -> B: pass
@overload
def f(x: C) -> C: pass


-- Default argument values
-- -----------------------


[case testCallingFunctionsWithDefaultArgumentValues]
from typing import Undefined
a, b = Undefined, Undefined # type: (A, B)
a = f()     # E: Incompatible types in assignment (expression has type "B", variable has type "A")
b = f(b)    # E: Argument 1 to "f" has incompatible type "B"; expected "A"
b = f(a, a) # E: Too many arguments for "f"

b = f()
b = f(a)
b = f(AA())

def f(x: 'A'  =  None) -> 'B': pass

class A: pass
class AA(A): pass
class B: pass

[case testDefaultArgumentExpressions]
import typing
def f(x: 'A' = A()) -> None:
    b = x # type: B # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    a = x # type: A

class B: pass
class A: pass
[out]
main: In function "f":

[case testDefaultArgumentExpressions2]
import typing
def f(x: 'A' = B()) -> None: # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    b = x # type: B      # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    a = x # type: A

class B: pass
class A: pass
[out]
main: In function "f":

[case testDefaultArgumentsWithSubtypes]
import typing
def f(x: 'B' = A()) -> None: # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    pass
def g(x: 'A' = B()) -> None:
    pass

class A: pass
class B(A): pass
[out]
main: In function "f":

[case testMultipleDefaultArgumentExpressions]
import typing
def f(x: 'A' = B(), y: 'B' = B()) -> None: # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    pass
def h(x: 'A' = A(), y: 'B' = B()) -> None:
    pass

class A: pass
class B: pass
[out]
main: In function "f":

[case testMultipleDefaultArgumentExpressions2]
import typing
def g(x: 'A' = A(), y: 'B' = A()) -> None: # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    pass

class A: pass
class B: pass
[out]
main: In function "g":

[case testDefaultArgumentsAndSignatureAsComment]
import typing
def f(x = 1): # type: (int) -> str
    pass
f()
f(1)
f('') # E: Argument 1 to "f" has incompatible type "str"; expected "int"

[case testMethodDefaultArgumentsAndSignatureAsComment]
import typing
class A:
    def f(self, x = 1): # type: (int) -> str
        pass
A().f()
A().f(1)
A().f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"


-- Access to method defined as a data attribute
-- --------------------------------------------


[case testMethodAsDataAttribute]
from typing import Undefined, Any, Callable
class B: pass
x = Undefined # type: Any
class A:
    f = x # type: Callable[[A], None]
    g = x # type: Callable[[A, B], None]
a = Undefined # type: A
a.f()
a.g(B())
a.f(a) # E: Too many arguments
a.g()  # E: Too few arguments

[case testMethodWithInvalidMethodAsDataAttribute]
from typing import Undefined, Any, Callable
class B: pass
x = Undefined # type: Any
class A:
    f = x # type: Callable[[], None]
    g = x # type: Callable[[B], None]
a = Undefined # type: A
a.f() # E: Invalid method type
a.g() # E: Invalid method type

[case testMethodWithDynamicallyTypedMethodAsDataAttribute]
from typing import Undefined, Any, Callable
class B: pass
x = Undefined # type: Any
class A:
    f = x # type: Callable[[Any], Any]
a = Undefined # type: A
a.f()
a.f(a) # E: Too many arguments

[case testOverloadedMethodAsDataAttribute]
from typing import overload, Undefined
class B: pass
class A:
    @overload
    def f(self) -> None: pass
    @overload
    def f(self, b: B) -> None: pass
    g = f
a = Undefined # type: A
a.g()
a.g(B())
a.g(a) # E: No overload variant matches argument types

[case testMethodAsDataAttributeInferredFromDynamicallyTypedMethod]
from typing import Undefined
class A:
    def f(self, x): pass
    g = f
a = Undefined # type: A
a.g(object())
a.g(a, a) # E: Too many arguments
a.g()     # E: Too few arguments

[case testMethodAsDataAttributeInGenericClass]
from typing import TypeVar, Generic, Undefined
t = TypeVar('t')
class B: pass
class A(Generic[t]):
    def f(self, x: t) -> None: pass
    g = f
a = Undefined # type: A[B]
a.g(B())
a.g(a)   # E: Argument 1 has incompatible type A[B]; expected "B"

[case testInvalidMethodAsDataAttributeInGenericClass]
from typing import Undefined, Any, TypeVar, Generic, Callable
t = TypeVar('t')
class B: pass
class C: pass
x = Undefined # type: Any
class A(Generic[t]):
    f = x # type: Callable[[A[B]], None]
ab = Undefined # type: A[B]
ac = Undefined # type: A[C]
ab.f()
ac.f()   # E: Invalid method type

[case testPartiallyTypedSelfInMethodDataAttribute]
from typing import Undefined, Any, TypeVar, Generic, Callable
t = TypeVar('t')
class B: pass
class C: pass
x = Undefined # type: Any
class A(Generic[t]):
    f = x # type: Callable[[A], None]
ab = Undefined # type: A[B]
ac = Undefined # type: A[C]
ab.f()
ac.f()

[case testCallableDataAttribute]
from typing import Undefined, Callable
class A:
    g = Undefined # type: Callable[[A], None]
    def __init__(self, f: Callable[[], None]) -> None:
        self.f = f
a = A(None)
a.f()
a.g()
a.f(a) # E: Too many arguments
a.g(a) # E: Too many arguments


-- Nested functions
-- ----------------


[case testSimpleNestedFunction]
import typing
def f(a: 'A') -> None:
    def g(b: 'B') -> None:
         b = a # fail
         aa = a # type: A # ok
         b = B()
    g(a) # fail
    g(B())
class A: pass
class B: pass
[out]
main: In function "g":
main, line 4: Incompatible types in assignment (expression has type "A", variable has type "B")
main: In function "f":
main, line 7: Argument 1 to "g" has incompatible type "A"; expected "B"

[case testReturnAndNestedFunction]
import typing
def f() -> 'A':
    def g() -> 'B':
        return A() # fail
        return B()
    return B() # fail
    return A()
class A: pass
class B: pass
[out]
main: In function "g":
main, line 4: Incompatible return value type: expected __main__.B, got __main__.A
main: In function "f":
main, line 6: Incompatible return value type: expected __main__.A, got __main__.B

[case testDynamicallyTypedNestedFunction]
import typing
def f(x: object) -> None:
    def g(y):
        pass
    g() # E: Too few arguments for "g"
    g(x)
[out]
main: In function "f":

[case testNestedFunctionInMethod]
import typing
class A:
    def f(self) -> None:
        def g(x: int) -> None:
            y = x # type: int
            a = x # type: A # fail
        g(2)
        g(A()) # fail
[out]
main: In function "g":
main, line 6: Incompatible types in assignment (expression has type "int", variable has type "A")
main: In member "f" of class "A":
main, line 8: Argument 1 to "g" has incompatible type "A"; expected "int"


-- Casts
-- -----


[case testCastsToAndFromFunctionTypes]
from typing import TypeVar, Callable, Any, cast
t = TypeVar('t')
def f(x: t,
      f1: Callable[[], None],
      f2: Callable[[Any], None], o: object) -> None:
    x = cast(t, f1)
    f1 = cast(Callable[[], None], x)
    f1 = cast(Callable[[], None], f2)
    f1 = cast(Callable[[], None], o)


-- Function decorators
-- -------------------


[case testTrivialStaticallyTypedFunctionDecorator]
from typing import TypeVar
t = TypeVar('t')
def dec(f: t) -> t:
    return f
@dec
def f(x: int) -> None: pass
f(1)
f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"

[case testTrivialStaticallyTypedMethodDecorator]
from typing import TypeVar
t = TypeVar('t')
def dec(f: t) -> t:
    return f
class A:
    @dec
    def f(self, x: int) -> None: pass
A().f(1)
A().f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"
class B: pass

[case testTrivialDecoratedNestedFunction]
from typing import TypeVar
t = TypeVar('t')
def dec(f: t) -> t:
    return f
def g() -> None:
    @dec
    def f(x: int) -> None: pass
    f(1)
    f('') # E: Argument 1 to "f" has incompatible type "str"; expected "int"
[out]
main: In function "g":

[case testCheckingDecoratedFunction]
import typing
def dec(f): pass
@dec
def f(x: 'A') -> None:
    a = x # type: A
    x = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
class A: pass
[out]
main: In function "f":

[case testDecoratorThatSwitchesType]
from typing import Callable
def dec(x) -> Callable[[], None]: pass
@dec
def f(y): pass
f()
f(None) # E: Too many arguments for "f"

[case testDecoratorThatSwitchesTypeWithMethod]
from typing import Any, Callable, Undefined
def dec(x) -> Callable[[Any], None]: pass
class A:
    @dec
    def f(self, a, b, c): pass
a = Undefined # type: A
a.f()
a.f(None) # E: Too many arguments for "f" of "A"

[case testNestedDecorators]
from typing import Any, Callable
def dec1(f: Callable[[Any], None]) -> Callable[[], None]: pass
def dec2(f: Callable[[Any, Any], None]) -> Callable[[Any], None]: pass
@dec1
@dec2
def f(x, y): pass
f()
f(None) # E: Too many arguments for "f"

[case testInvalidDecorator1]
from typing import Any, Callable
def dec1(f: Callable[[Any], None]) -> Callable[[], None]: pass
def dec2(f: Callable[[Any, Any], None]) -> Callable[[Any], None]: pass
@dec1 # E: Argument 1 to "dec2" has incompatible type Callable[[Any], Any]; expected Callable[[Any, Any], None]
@dec2
def f(x): pass

[case testInvalidDecorator2]
from typing import Any, Callable
def dec1(f: Callable[[Any, Any], None]) -> Callable[[], None]: pass
def dec2(f: Callable[[Any, Any], None]) -> Callable[[Any], None]: pass
@dec1 # E: Argument 1 to "dec1" has incompatible type Callable[[Any], None]; expected Callable[[Any, Any], None]
@dec2
def f(x, y): pass


-- Conditional function definition
-- -------------------------------


[case testTypeCheckBodyOfConditionalFunction]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(x: int) -> None:
        x = 1
        x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
[out]
main: In function "f":

[case testCallConditionalFunction]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(x: int) -> None:  pass
    f(1)
    f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"
f(1)
f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"

[case testConditionalFunctionDefinitionWithIfElse]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(x: int) -> None:
        x = 'x'   # fail
        x = 1
else:
    def f(x: int) -> None:
        x + 'x'   # fail
        x = 1
f(1)
f('x') # fail
[out]
main: In function "f":
main, line 5: Incompatible types in assignment (expression has type "str", variable has type "int")
main, line 9: Unsupported left operand type for + ("int")
main: At top level:
main, line 12: Argument 1 to "f" has incompatible type "str"; expected "int"

[case testUnconditionalRedefinitionOfConditionalFunction]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(): pass
def f(): pass # E: Name 'f' already defined (overload variants must be next to each other)

[case testIncompatibleConditionalFunctionDefinition]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(x: int) -> None: pass
else:
    def f(x): pass # E: All conditional function variants must have identical signatures

[case testIncompatibleConditionalFunctionDefinition2]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(x: int) -> None: pass
else:
    def f(y: int) -> None: pass # E: All conditional function variants must have identical signatures

[case testIncompatibleConditionalFunctionDefinition3]
from typing import Undefined, Any
x = Undefined # type: Any
if x:
    def f(x: int) -> None: pass
else:
    def f(x: int = 0) -> None: pass # E: All conditional function variants must have identical signatures


-- Conditional method definition
-- -----------------------------


[case testTypeCheckBodyOfConditionalMethod]
from typing import Undefined, Any
x = Undefined # type: Any
class A:
    if x:
        def f(self, x: int) -> None:
            x = 1
            x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
[out]
main: In member "f" of class "A":

[case testCallConditionalMethodInClassBody]
from typing import Undefined, Any
x = Undefined # type: Any
class A:
    if x:
        def f(self, x: int) -> None:  pass
        f(x, 1)
        f(x, 'x') # E: Argument 2 to "f" of "A" has incompatible type "str"; expected "int"
    f(x, 1)
    f(x, 'x') # E: Argument 2 to "f" of "A" has incompatible type "str"; expected "int"
[out]
main: In class "A":

[case testCallConditionalMethodViaInstance]
from typing import Undefined, Any
x = Undefined # type: Any
class A:
    if x:
         def f(self, x: int) -> None: pass
A().f(1)
A().f('x') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"

[case testConditionalMethodDefinitionWithIfElse]
from typing import Undefined, Any
x = Undefined # type: Any
class A:
    if x:
        def f(self, x: int) -> None:
            x = 'x'   # fail
            x = 1
    else:
        def f(self, x: int) -> None:
            x + 'x'   # fail
            x = 1
A().f(1)
A().f('x') # fail
[out]
main: In member "f" of class "A":
main, line 6: Incompatible types in assignment (expression has type "str", variable has type "int")
main, line 10: Unsupported left operand type for + ("int")
main: At top level:
main, line 13: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"

[case testUnconditionalRedefinitionOfConditionalMethod]
from typing import Undefined, Any
x = Undefined # type: Any
class A:
    if x:
        def f(self): pass
    def f(self): pass # E: Name 'f' already defined

[case testIncompatibleConditionalMethodDefinition]
from typing import Undefined, Any
x = Undefined # type: Any
class A:
    if x:
        def f(self, x: int) -> None: pass
    else:
        def f(self, x): pass # E: All conditional function variants must have identical signatures
[out]
main: In class "A":

[case testConditionalFunctionDefinitionInTry]
import typing
try:
    def f(x: int) -> None: pass
except:
    def g(x: str) -> None: pass
f(1)
f('x') # E: Argument 1 to "f" has incompatible type "str"; expected "int"
g('x')
g(1) # E: Argument 1 to "g" has incompatible type "int"; expected "str"
