-- Methods
-- -------


[case testMethodCall]
from typing import Undefined
a = Undefined # type: A
b = Undefined # type: B

a.foo(B())        # Fail
a.bar(B(), A())   # Fail

a.foo(A())
b.bar(B(), A())

class A:
    def foo(self, x: 'A') -> None: pass
class B:
    def bar(self, x: 'B', y: A) -> None: pass
[out]
main, line 5: Argument 1 to "foo" of "A" has incompatible type "B"; expected "A"
main, line 6: "A" has no attribute "bar"

[case testMethodCallWithSubtype]
from typing import Undefined
a = Undefined # type: A
a.foo(A())
a.foo(B())
a.bar(A()) # Fail
a.bar(B())

class A:
    def foo(self, x: 'A') -> None: pass
    def bar(self, x: 'B') -> None: pass
class B(A): pass
[out]
main, line 5: Argument 1 to "bar" of "A" has incompatible type "A"; expected "B"

[case testInheritingMethod]
from typing import Undefined
a = Undefined # type: B
a.foo(A()) # Fail
a.foo(B())

class A:
    def foo(self, x: 'B') -> None: pass
class B(A): pass
[out]
main, line 3: Argument 1 to "foo" of "A" has incompatible type "A"; expected "B"

[case testMethodCallWithInvalidNumberOfArguments]
from typing import Undefined
a = Undefined # type: A
a.foo()               # Fail
a.foo(object(), A())  # Fail

class A:
    def foo(self, x: 'A') -> None: pass
[out]
main, line 3: Too few arguments for "foo" of "A"
main, line 4: Too many arguments for "foo" of "A"
main, line 4: Argument 1 to "foo" of "A" has incompatible type "object"; expected "A"

[case testMethodBody]
import typing
class A:
    def f(self) -> None:
        a = object() # type: A    # Fail
[out]
main: In member "f" of class "A":
main, line 4: Incompatible types in assignment (expression has type "object", variable has type "A")

[case testMethodArguments]
import typing
class A:
    def f(self, a: 'A', b: 'B') -> None:
        a = B() # Fail
        b = A() # Fail
        a = A()
        b = B()
        a = a
        a = b # Fail
class B: pass
[out]
main: In member "f" of class "A":
main, line 4: Incompatible types in assignment (expression has type "B", variable has type "A")
main, line 5: Incompatible types in assignment (expression has type "A", variable has type "B")
main, line 9: Incompatible types in assignment (expression has type "B", variable has type "A")

[case testReturnFromMethod]
import typing
class A:
    def f(self) -> 'A':
        return B() # Fail
        return A()
class B: pass
[out]
main: In member "f" of class "A":
main, line 4: Incompatible return value type: expected __main__.A, got __main__.B

[case testSelfArgument]
import typing
class A:
    def f(self) -> None:
        o = self # type: B    # Fail
        self.g()      # Fail
        a = self # type: A
        self.f()
class B: pass
[out]
main: In member "f" of class "A":
main, line 4: Incompatible types in assignment (expression has type "A", variable has type "B")
main, line 5: "A" has no attribute "g"

[case testAssignToMethodViaInstance]
import typing
class A:
    def f(self): pass
A().f = None # E: Cannot assign to a method


-- Attributes
-- ----------


[case testReferToInvalidAttribute]
from typing import Undefined
class A:
    def __init__(self):
        self.x = object()
a = Undefined # type: A
a.y
a.y = object()
a.x
a.x = object()
[out]
main, line 6: "A" has no attribute "y"
main, line 7: "A" has no attribute "y"

[case testArgumentTypeInference]
from typing import Undefined
class A:
    def __init__(self, aa: 'A', bb: 'B') -> None:
        self.a = aa
        self.b = bb
class B: pass
a = Undefined # type: A
b = Undefined # type: B
a.a = b # Fail
a.b = a # Fail
b.a     # Fail
a.a = a
a.b = b
[out]
main, line 9: Incompatible types in assignment (expression has type "B", variable has type "A")
main, line 10: Incompatible types in assignment (expression has type "A", variable has type "B")
main, line 11: "B" has no attribute "a"

[case testExplicitAttributeInBody]
from typing import Undefined
a = Undefined # type: A
a.x = object() # Fail
a.x = A()
class A:
  x = Undefined # type: A
[out]
main, line 3: Incompatible types in assignment (expression has type "object", variable has type "A")

[case testAttributeDefinedInNonInitMethod]
import typing
class A:
    def f(self) -> None:
        self.x = 1
        self.y = ''
        self.x = 1
a = A()
a.x = 1
a.y = ''
a.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
a.z = 0  # E: "A" has no attribute "z"

[case testInheritanceAndAttributeAssignment]
import typing
class A:
    def f(self) -> None:
        self.x = 0
class B(A):
    def f(self) -> None:
        self.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
[out]
main: In member "f" of class "B":

[case testAssignmentToAttributeInMultipleMethods]
import typing
class A:
    def f(self) -> None:
        self.x = 0
    def g(self) -> None:
        self.x = '' # Fail
    def __init__(self) -> None:
        self.x = '' # Fail
[out]
main: In member "g" of class "A":
main, line 6: Incompatible types in assignment (expression has type "str", variable has type "int")
main: In member "__init__" of class "A":
main, line 8: Incompatible types in assignment (expression has type "str", variable has type "int")


-- Method overriding
-- -----------------


[case testMethodOverridingWithIdenticalSignature]
import typing
class A:
    def f(self, x: 'A') -> None: pass
    def g(self, x: 'B' , y: object) -> 'A': pass
    def h(self) -> None: pass
class B(A):
    def f(self, x: A) -> None: pass
    def g(self, x: 'B' , y: object) -> A: pass
    def h(self) -> None: pass
[out]

[case testMethodOverridingWithReturnTypeCoVariance]
import typing
class A:
  def f(self, x: 'A', y: 'B') -> 'A': pass
class B(A):
  def f(self, x: A, y: 'B') -> 'B': pass
[out]

[case testMethodOverridingWithIncompatibleTypes]
import typing
class A:
  def f(self, x: 'A', y: 'B') -> 'A': pass
  def g(self, x: 'A', y: 'B') -> 'A': pass
  def h(self, x: 'A', y: 'B') -> 'A': pass
class B(A):
  def f(self, x: 'B', y: 'B') -> A: pass  # Fail
  def g(self, x: A, y: A) -> A: pass
  def h(self, x: A, y: 'B') -> object: pass  # Fail
[out]
main: In class "B":
main, line 7: Argument 1 of "f" incompatible with supertype "A"
main, line 9: Return type of "h" incompatible with supertype "A"

[case testMethodOverridingWithIncompatibleArgumentCount]
import typing
class A:
    def f(self, x: 'A') -> None: pass
    def g(self, x: 'A', y: 'B') -> 'A': pass
class B(A):
    def f(self, x: A, y: A) -> None: pass # Fail
    def g(self, x: A) -> A: pass # Fail
[out]
main: In class "B":
main, line 6: Signature of "f" incompatible with supertype "A"
main, line 7: Signature of "g" incompatible with supertype "A"

[case testMethodOverridingAcrossDeepInheritanceHierarchy1]
import typing
class A:
    def f(self, x: 'B') -> None: pass
class B(A): pass
class C(B): # with gap in implementations
    def f(self, x: 'C') -> None:  # Fail
        pass
[out]
main: In class "C":
main, line 6: Argument 1 of "f" incompatible with supertype "A"

[case testMethodOverridingAcrossDeepInheritanceHierarchy2]
import typing
class A:
    def f(self) -> 'B': pass
class B(A):
    def f(self) -> 'C': pass
class C(B): # with multiple implementations
    def f(self) -> B:  # Fail
        pass
[out]
main: In class "C":
main, line 7: Return type of "f" incompatible with supertype "B"

[case testMethodOverridingWithVoidReturnValue]
import typing
class A:
    def f(self) -> None: pass
    def g(self) -> 'A': pass
class B(A):
    def f(self) -> A: pass  # Fail
    def g(self) -> None: pass  # Fail
[out]
main: In class "B":
main, line 6: Return type of "f" incompatible with supertype "A"
main, line 7: Return type of "g" incompatible with supertype "A"


-- Constructors
-- ------------


[case testTrivialConstructor]
import typing
a = A() # type: A
b = A() # type: B # Fail
class A:
    def __init__(self) -> None: pass
class B: pass
[out]
main, line 3: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testConstructor]
import typing
a = A(B()) # type: A
aa = A(object()) # type: A  # Fail
b = A(B()) # type: B       # Fail
class A:
    def __init__(self, x: 'B') -> None: pass
class B: pass
[out]
main, line 3: Argument 1 to "A" has incompatible type "object"; expected "B"
main, line 4: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testConstructorWithTwoArguments]
import typing
a = A(C(), B()) # type: A  # Fail

class A:
    def __init__(self, x: 'B', y: 'C') -> None: pass
class B: pass
class C(B): pass
[out]
main, line 2: Argument 2 to "A" has incompatible type "B"; expected "C"

[case testInheritedConstructor]
import typing
b = B(C()) # type: B
a = B(D()) # type: A # Fail
class A:
    def __init__(self, x: 'C') -> None: pass
class B(A): pass
class C: pass
class D: pass
[out]
main, line 3: Argument 1 to "B" has incompatible type "D"; expected "C"

[case testOverridingWithIncompatibleConstructor]
import typing
A()    # Fail
B(C()) # Fail
A(C())
B()
class A:
    def __init__(self, x: 'C') -> None: pass
class B(A):
    def __init__(self) -> None: pass
class C: pass
[out]
main, line 2: Too few arguments for "A"
main, line 3: Too many arguments for "B"

[case testConstructorWithReturnValueType]
import typing
class A:
    def __init__(self) -> 'A': pass
[out]
main: In member "__init__" of class "A":
main, line 3: Cannot define return type for "__init__"

[case testGlobalFunctionInitWithReturnType]
import typing
a = __init__() # type: A
b = __init__() # type: B # Fail
def __init__() -> 'A': pass
class A: pass
class B: pass
[out]
main, line 3: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testAccessingInit]
from typing import Undefined, Any
class A:
    def __init__(self, a: 'A') -> None: pass
a = Undefined # type: A
a.__init__(a)  # E: Cannot access "__init__" directly
(Any(a)).__init__(a)

[case testDeepInheritanceHierarchy]
import typing
d = C() # type: D  # Fail
d = B()      # Fail
d = A()      # Fail
d = D2()     # Fail
a = D() # type: A
a = D2()
b = D() # type: B
b = D2()

class A: pass
class B(A): pass
class C(B): pass
class D(C): pass
class D2(C): pass
[out]
main, line 2: Incompatible types in assignment (expression has type "C", variable has type "D")
main, line 3: Incompatible types in assignment (expression has type "B", variable has type "D")
main, line 4: Incompatible types in assignment (expression has type "A", variable has type "D")
main, line 5: Incompatible types in assignment (expression has type "D2", variable has type "D")


-- Attribute access in class body
-- ------------------------------


[case testDataAttributeRefInClassBody]
import typing
class B: pass
class A:
    x = B()
    y = x
    b = x # type: B
    b = x
    c = x # type: A # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    c = b   # E: Incompatible types in assignment (expression has type "B", variable has type "A")
[out]
main: In class "A":

[case testMethodRefInClassBody]
from typing import Function
class B: pass
class A:
    def f(self) -> None: pass
    g = f
    h = f # type: Function[[A], None]
    h = f
    g = h
    ff = f # type: Function[[B], None]  # E: Incompatible types in assignment (expression has type Function[[A], None], variable has type Function[[B], None])
    g = ff                # E: Incompatible types in assignment (expression has type Function[[B], None], variable has type Function[[A], None])
[out]
main: In class "A":


-- Arbitrary statements in class body
-- ----------------------------------


[case testStatementsInClassBody]
import typing
class B: pass
class A:
    for x in [A()]:
        y = x
        y = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
    x = A()
    y = A()
    x = B() # E: Incompatible types in assignment (expression has type "B", variable has type "A")
[builtins fixtures/for.py]
[out]
main: In class "A":


-- Class attributes
-- ----------------


[case testAccessMethodViaClass]
import typing
class A:
    def f(self) -> None: pass
A.f(A())
A.f(object())     # E: Argument 1 to "f" of "A" has incompatible type "object"; expected "A"
A.f()             # E: Too few arguments for "f" of "A"
A.f(None, None)   # E: Too many arguments for "f" of "A"

[case testAccessAttributeViaClass]
import typing
class B: pass
class A:
    x = None # type: A
a = A.x # type: A
b = A.x # type: B # E: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testAccessingUndefinedAttributeViaClass]
import typing
class A: pass
A.x # E: "A" has no attribute "x"

[case testAccessingUndefinedAttributeViaClassWithOverloadedInit]
from typing import overload
class A:
    @overload
    def __init__(self): pass
    @overload
    def __init__(self, x): pass
A.x # E: "A" has no attribute "x"

[case testAccessMethodOfClassWithOverloadedInit]
from typing import overload, Any
class A:
    @overload
    def __init__(self) -> None: pass
    @overload
    def __init__(self, x: Any) -> None: pass
    def f(self) -> None: pass
A.f(A())
A.f()    # E: Too few arguments for "f" of "A"

[case testAssignmentToClassDataAttribute]
import typing
class B: pass
class A:
    x = None # type: B
A.x = B()
A.x = object()  # E: Incompatible types in assignment (expression has type "object", variable has type "B")

[case testAssignmentToInferredClassDataAttribute]
import typing
class B: pass
class A:
     x = B()
A.x = B()
A.x = A()   # E: Incompatible types in assignment (expression has type "A", variable has type "B")

[case testInitMethodUnbound]
from typing import Undefined
class B: pass
class A:
    def __init__(self, b: B) -> None: pass
a = Undefined # type: A
b = Undefined # type: B
A.__init__(a, b)
A.__init__(b, b) # E: Argument 1 to "__init__" of "A" has incompatible type "B"; expected "A"
A.__init__(a, a) # E: Argument 2 to "__init__" of "A" has incompatible type "A"; expected "B"

[case testAssignToMethodViaClass]
import typing
class A:
    def f(self): pass
A.f = None # E: Cannot assign to a method

[case testAssignToNestedClassViaClass]
import typing
class A:
    class B: pass
A.B = None # E: Cannot assign to a type


-- Nested classes
-- --------------


[case testClassWithinFunction]
from typing import Undefined
def f() -> None:
    class A:
        def g(self) -> None: pass
    a = Undefined # type: A
    a.g()
    a.g(a) # E: Too many arguments for "g" of "A"
[out]
main: In function "f":

[case testConstructNestedClass]
import typing
class A:
    class B: pass
    b = B()
    b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")
    b = B(b) # E: Too many arguments for "B"
[out]
main: In class "A":

[case testConstructNestedClassWithCustomInit]
import typing
class A:
    def f(self) -> None:
        class B:
            def __init__(self, a: 'A') -> None: pass
        b = B(A())
        b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")
        b = B() # E: Too few arguments for "B"
[out]
main: In member "f" of class "A":

[case testDeclareVariableWithNestedClassType]
from typing import Undefined
def f() -> None:
    class A: pass
    a = Undefined # type: A
    a = A()
    a = object() # E: Incompatible types in assignment (expression has type "object", variable has type "A")
[out]
main: In function "f":

[case testExternalReferenceToClassWithinClass]
from typing import Undefined
class A:
    class B: pass
b = Undefined # type: A.B
b = A.B()
b = A() # E: Incompatible types in assignment (expression has type "A", variable has type "B")
b = A.B(b) # E: Too many arguments for "B"


-- Declaring attribute type in method
-- ----------------------------------


[case testDeclareAttributeTypeInInit]
from typing import Undefined
class A:
    def __init__(self):
        self.x = None # type: int
a = Undefined # type: A
a.x = 1
a.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")

[case testAccessAttributeDeclaredInInitBeforeDeclaration]
from typing import Undefined
a = Undefined # type: A
a.x = 1
a.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
class A:
    def __init__(self):
        self.x = None # type: int


-- Special cases
-- -------------


[case testMultipleClassDefinition]
import typing
A()
class A: pass
class A: pass
[out]
main, line 4: Name 'A' already defined

[case testDocstringInClass]
import typing
class A:
    """Foo"""
class B:
    'x'
    y = B()
[builtins fixtures/primitives.py]

[case testErrorMessageInFunctionNestedWithinMethod]
import typing
class A:
    def f(self) -> None:
        def g() -> None:
            a = None
        b = None
[out]
main: In function "g":
main, line 5: Need type annotation for variable
main: In member "f" of class "A":
main, line 6: Need type annotation for variable


-- Using Undefined(type) to declare attributes
-- -------------------------------------------


[case testDeclareClassVariableUsingUndefinedWithArgument]
from typing import Undefined
class A:
    x = Undefined(int)
a = Undefined(A)
a.x = 1
a.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")

[case testDeclareInstanceVariableUsingUndefinedWithArgument]
from typing import Undefined
class A:
    def __init__(self):
        self.x = Undefined(int)
a = Undefined(A)
a.x = 1
a.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")

[case testAccessInstanceVariableDeclaredUsingUndefinedWithArgumentBeforeDecl]
from typing import Undefined
class A:
    def f(self) -> None:
        self.x = 1
        self.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
    def __init__(self):
        self.x = Undefined(int)
[out]
main: In member "f" of class "A":


-- Static methods
-- --------------


[case testSimpleStaticMethod]
import typing
class A:
  @staticmethod
  def f(x: int) -> None: pass
A.f(1)
A().f(1)
A.f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"
A().f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"
[builtins fixtures/staticmethod.py]

[case testBuiltinStaticMethod]
import typing
int.from_bytes(b'', '')
int.from_bytes('', '') # E: Argument 1 to "from_bytes" of "int" has incompatible type "str"; expected "bytes"
[builtins fixtures/staticmethod.py]

[case testAssignStaticMethodOnInstance]
import typing
class A:
  @staticmethod
  def f(x: int) -> None: pass
A().f = A.f # E: Cannot assign to a method
[builtins fixtures/staticmethod.py]


-- Class methods
-- -------------


[case testSimpleClassMethod]
import typing
class A:
  @classmethod
  def f(cls, x: int) -> None: pass
A.f(1)
A().f(1)
A.f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"
A().f('') # E: Argument 1 to "f" of "A" has incompatible type "str"; expected "int"
[builtins fixtures/classmethod.py]

[case testBuiltinClassMethod]
import typing
int.from_bytes(b'', '')
int.from_bytes('', '') # E: Argument 1 to "from_bytes" of "int" has incompatible type "str"; expected "bytes"
[builtins fixtures/classmethod.py]

[case testAssignClassMethodOnClass]
import typing
class A:
  @classmethod
  def f(cls, x: int) -> None: pass
A.f = A.f # E: Cannot assign to a method
[builtins fixtures/classmethod.py]

[case testAssignClassMethodOnInstance]
import typing
class A:
  @classmethod
  def f(cls, x: int) -> None: pass
A().f = A.f # E: Cannot assign to a method
[builtins fixtures/classmethod.py]


-- Properties
-- ----------


[case testAccessingReadOnlyProperty]
import typing
class A:
    @property
    def f(self) -> str: pass
a = A() # type. A
s = ''  # type: str
s = a.f
a = a.f  # E: Incompatible types in assignment (expression has type "str", variable has type "A")
[builtins fixtures/property.py]

[case testAssigningToReadOnlyProperty]
import typing
class A:
    @property
    def f(self) -> str: pass
A().f = '' # E: Property "f" defined in "A" is read-only
[builtins fixtures/property.py]

[case testPropertyGetterBody]
import typing
class A:
    @property
    def f(self) -> str:
        self.x = 1
        self.x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
[builtins fixtures/property.py]
[out]
main: In member "f" of class "A":

[case testDynamicallyTypedProperty]
import typing
class A:
    @property
    def f(self): pass
a = A() # type. A
a.f.xx
a.f = '' # E: Property "f" defined in "A" is read-only
[builtins fixtures/property.py]


-- Multiple inheritance, non-object built-in class as base
-- -------------------------------------------------------


[case testInvalidMultipleInheritanceFromBuiltins]
import typing
class A(int, str): pass # E: Instance layout conflict in multiple inheritance
[out]
main: In class "A":

[case testInvalidMultipleInheritanceFromBuiltins]
import typing
class S(str): pass
class A(S, str): pass
class B(S, int): pass # E: Instance layout conflict in multiple inheritance
[out]
main: In class "B":


-- Ducktype decorators
-- -------------------


[case testSimpleDucktypeDecorator]
from typing import ducktype, Undefined
class A: pass
@ducktype(A)
class B: pass
a = Undefined(A)
b = Undefined(B)
b = a # E: Incompatible types in assignment (expression has type "A", variable has type "B")
a = b

[case testDucktypeTransitivityDecorator]
from typing import ducktype, Undefined
class A: pass
@ducktype(A)
class B: pass
@ducktype(B)
class C: pass
a = Undefined(A)
c = Undefined(C)
c = a # E: Incompatible types in assignment (expression has type "A", variable has type "C")
a = c


-- Disjointclass decorators
-- ------------------------


[case testDisjointclassDecorator]
from typing import disjointclass
class A: pass
@disjointclass(A)
class B: pass
class C(A, B): pass # Error
class D(B, A): pass # Error
b = B()
[out]
main: In class "C":
main, line 5: disjointclass constraint of class B disallows A as a base class
main: In class "D":
main, line 6: disjointclass constraint of class B disallows A as a base class


-- Operator methods
-- ----------------


[case testOperatorMethodOverrideIntroducingOverloading]
from typing import overload
class A:
    def __add__(self, x: int) -> int: pass
class B(A):
    @overload  # E: Signature of "__add__" incompatible with supertype "A"
    def __add__(self, x: int) -> int: pass
    @overload
    def __add__(self, x: str) -> str: pass
[out]
main: In class "B":

[case testOperatorMethodOverrideWideningArgumentType]
import typing
class A:
    def __add__(self, x: int) -> int: pass
class B(A):
    def __add__(self, x: object) -> int: pass
[out]

[case testOperatorMethodOverrideNarrowingReturnType]
import typing
class A:
    def __add__(self, x: int) -> 'A': pass
class B(A):
    def __add__(self, x: int) -> 'B': pass

[case testOperatorMethodOverrideWithDynamicallyTyped]
import typing
class A:
    def __add__(self, x: int) -> 'A': pass
class B(A):
    def __add__(self, x): pass

[case testOperatorMethodOverrideWithIdenticalOverloadedType]
from typing import overload
class A:
    @overload
    def __add__(self, x: int) -> 'A': pass
    @overload
    def __add__(self, x: str) -> 'A': pass
class B(A):
    @overload
    def __add__(self, x: int) -> 'A': pass
    @overload
    def __add__(self, x: str) -> 'A': pass

[case testOverloadedOperatorMethodOverrideWithDynamicallyTypedMethod]
from typing import overload, Any
class A:
    @overload
    def __add__(self, x: int) -> 'A': pass
    @overload
    def __add__(self, x: str) -> 'A': pass
class B(A):
    def __add__(self, x): pass
class C(A):
    def __add__(self, x: Any) -> A: pass

[case testOverloadedOperatorMethodOverrideWithNewItem]
from typing import overload, Any
class A:
    @overload
    def __add__(self, x: int) -> 'A': pass
    @overload
    def __add__(self, x: str) -> 'A': pass
class B(A):
    @overload
    def __add__(self, x: int) -> A: pass
    @overload
    def __add__(self, x: str) -> A: pass
    @overload
    def __add__(self, x: type) -> A: pass
[out]
main: In class "B":
main, line 8: Signature of "__add__" incompatible with supertype "A"

[case testOverloadedOperatorMethodOverrideWithSwitchedItemOrder]
from typing import overload, Any
class A:
    @overload
    def __add__(self, x: 'B') -> 'B': pass
    @overload
    def __add__(self, x: 'A') -> 'A': pass
class B(A):
    @overload
    def __add__(self, x: 'A') -> 'A': pass
    @overload
    def __add__(self, x: 'B') -> 'B': pass
[out]
main: In class "B":
main, line 8: Signature of "__add__" incompatible with supertype "A"

[case testReverseOperatorMethoddArgumentType]
from typing import Any
class A: pass
class B:
    def __radd__(self, x: A) -> int: pass # Error
class C:
    def __radd__(self, x: A) -> Any: pass
class D:
    def __radd__(self, x: A) -> object: pass
[out]
main: In member "__radd__" of class "B":
main, line 4: "Any" return type expected since argument to __radd__ does not support __add__

[case testReverseOperatorMethoddArgumentType2]
from typing import Any, Tuple, Function
class A:
    def __radd__(self, x: Tuple[int, str]) -> int: pass
class B:
    def __radd__(self, x: Function[[], int]) -> int: pass
class C:
    def __radd__(self, x: Any) -> int: pass
[out]
main: In member "__radd__" of class "A":
main, line 3: "Any" return type expected since argument to __radd__ does not support __add__
main: In member "__radd__" of class "B":
main, line 5: "Any" return type expected since argument to __radd__ does not support __add__
main: In member "__radd__" of class "C":
main, line 7: "Any" return type expected since argument to __radd__ has type "Any"

[case testOverloadedReverseOperatorMethodArgumentType]
from typing import overload, Any
class A:
    @overload
    def __radd__(self, x: 'A') -> str: pass # Error
    @overload
    def __radd__(self, x: 'A') -> Any: pass
[out]
main: In member "__radd__" of class "A":
main, line 4: "Any" return type expected since argument to __radd__ does not support __add__

[case testReverseOperatorMethodArgumentTypeAndOverloadedMethod]
from typing import overload, disjointclass
@disjointclass(int)
@disjointclass(str)
class A:
    @overload
    def __add__(self, x: int) -> int: pass
    @overload
    def __add__(self, x: str) -> int: pass
    def __radd__(self, x: 'A') -> str: pass

[case testAbstractReverseOperatorMethod]
import typing
from abc import abstractmethod
class A:
    @abstractmethod
    def __lt__(self, x: 'A') -> int: pass
class B:
    @abstractmethod
    def __lt__(self, x: 'B') -> int: pass
    @abstractmethod
    def __gt__(self, x: 'B') -> int: pass
[out]
main: In member "__lt__" of class "A":
main, line 5: "Any" return type expected since argument to __lt__ does not support __gt__

[case testUnsafeOverlappingForwardAndReverseMethods]
import typing
class A:
    def __add__(self, x: int) -> str: pass
class B:
    def __radd__(self, x: A) -> int: pass # Error
[out]
main: In member "__radd__" of class "B":
main, line 5: Signatures of "__radd__" of "B" and "__add__" of "A" are unsafely overlapping

[case testUnsafeOverlappingWithOperatorMethodsAndOverloading]
from typing import overload
class A:
    @overload
    def __add__(self, x: 'A') -> int: pass
    @overload
    def __add__(self, x: str) -> int: pass
class B:
    def __radd__(self, x: 'A') -> str: pass # Error
[out]
main: In member "__radd__" of class "B":
main, line 8: Signatures of "__radd__" of "B" and "__add__" of "A" are unsafely overlapping

[case testUnsafeOverlappingWithOperatorMethodsAndOverloading2]
from typing import overload
class A:
    def __add__(self, x: 'A') -> int: pass
class B:
    @overload
    def __radd__(self, x: 'X') -> str: pass # Error
    @overload
    def __radd__(self, x: A) -> str: pass   # Error
class X:
    def __add__(self, x): pass
[out]
main: In member "__radd__" of class "B":
main, line 6: Signatures of "__radd__" of "B" and "__add__" of "X" are unsafely overlapping
main, line 8: Signatures of "__radd__" of "B" and "__add__" of "A" are unsafely overlapping

[case testOverlappingNormalAndInplaceOperatorMethod]
import typing
class A:
    # Incompatible (potential trouble with __radd__)
    def __add__(self, x: 'A') -> int: pass
    def __iadd__(self, x: 'B') -> int: pass
class B:
    # Safe
    def __add__(self, x: 'C') -> int: pass
    def __iadd__(self, x: A) -> int: pass
class C(A): pass
[out]
main: In class "A":
main, line 5: Signatures of "__iadd__" and "__add__" are incompatible

[case testOverloadedNormalAndInplaceOperatorMethod]
from typing import overload
class A:
    @overload
    def __add__(self, x: int) -> int: pass
    @overload
    def __add__(self, x: str) -> int: pass
    @overload # Error
    def __iadd__(self, x: int) -> int: pass
    @overload
    def __iadd__(self, x: object) -> int: pass
class B:
    @overload
    def __add__(self, x: int) -> int: pass
    @overload
    def __add__(self, x: str) -> str: pass
    @overload
    def __iadd__(self, x: int) -> int: pass
    @overload
    def __iadd__(self, x: str) -> str: pass
[out]
main: In class "A":
main, line 7: Signatures of "__iadd__" and "__add__" are incompatible

[case testIntroducingInplaceOperatorInSubclass]
import typing
class A:
    def __add__(self, x: 'A') -> 'B': pass
class B(A):
    # __iadd__ effectively partially overrides __add__
    def __iadd__(self, x: 'A') -> 'A': pass # Error
class C(A):
    def __iadd__(self, x: int) -> 'B': pass # Error
class D(A):
    def __iadd__(self, x: 'A') -> 'B': pass
[out]
main: In class "B":
main, line 6: Return type of "__iadd__" incompatible with "__add__" of supertype "A"
main: In class "C":
main, line 8: Argument 1 of "__iadd__" incompatible with "__add__" of supertype "A"
main, line 8: Signatures of "__iadd__" and "__add__" are incompatible


[case testGetAttr]
from typing import Undefined
a, b = Undefined, Undefined  # type: A, B
class A:
    def __getattr__(self, x: str) -> A:
        return A()
class B: pass

a = a.foo
b = a.bar
[out]
main, line 9: Incompatible types in assignment (expression has type "A", variable has type "B")


[case testGetAttrSignature]
class A:
    def __getattr__(self, x: str) -> A: pass
class B:
    def __getattr__(self, x: A) -> B: pass
class C:
    def __getattr__(self, x: str, y: str) -> C: pass
class D:
    def __getattr__(self, x: str) -> None: pass
[out]
main: In member "__getattr__" of class "B":
main, line 4: Invalid signature "def (self: __main__.B, x: __main__.A) -> __main__.B"
main: In member "__getattr__" of class "C":
main, line 6: Invalid signature "def (self: __main__.C, x: builtins.str, y: builtins.str) -> __main__.C"


-- Special cases
-- -------------


[case testSubclassInt]
import typing
class A(int): pass
n = 0
n = A()
a = A()
a = 0 # E: Incompatible types in assignment (expression has type "int", variable has type "A")


-- TODO
--   attribute inherited from superclass; assign in __init__
--   refer to attribute before type has been inferred (the initialization in
--   __init__ has not been analysed)
